File: py/algorithms/knuth_morris_pratt.py
""" This module contains code representing the Knuth-Morris-Pratt string
search algorithm """


def knuth_morris_pratt(bigstr: str, substr: str) -> int:
    """ This algorithm represents the Knuth-Morris-Pratt(KMP) string-search
    algorithm. This algorithm is used to search for occurrences of a word X
    within another word Y.

    Time:
        O(N+M) best/avg/worst

    Space:
        O(M) best/avg/worst

    Where N is the length of bigstr, M is the length of substr.

    Args:
        bigstr:
            String representing the string we are looking for the pattern substr
            within

        substr:
            String representing the string we are looking for within the bigstr

    Returns:
        Integer indicating where the substr starts within the bigstr. If not
        found, the function will return -1.

    """
    if not substr:
        return -1
    prefix_suffix_table = make_prefix_suffix_table(substr)
    return match(bigstr, substr, prefix_suffix_table)


def make_prefix_suffix_table(substr):
    table = [-1] * len(substr)
    i = 0
    j = 1
    while j < len(substr):
        if substr[j] == substr[i]:
            table[j] = i
            j += 1
            i += 1
        elif i != 0:
            i = table[i - 1] + 1
        else:
            j += 1
    return table


def match(bigstr, substr, prefix_suffix_table):
    i = 0
    j = 0
    while i < len(bigstr) and j < len(substr):
        if bigstr[i] == substr[j]:
            i += 1
            j += 1
        elif j != 0:
            j = prefix_suffix_table[j - 1] + 1
        else:
            i += 1
    return i - j if j == len(substr) else -1

File: ts/src/algorithms/knuth_morris_pratt.ts
function makePrefixSuffixTable(substr: string): number[] {
    // pointer i always moves forward
    let i = 1;
    // pointer j moves forward until match is not found
    // and then moves backward intelligently.
    let j = 0;

    const table = Array(substr.length).fill(-1) as number[];
    while (i < substr.length) {
        if (substr[i] === substr[j]) {
            table[i] = j;
            j += 1;
            i += 1;
        } else if (j !== 0) {
            j = table[j - 1] + 1;
        } else {
            i += 1;
        }
    }
    return table;
}

function match(
    bigstr: string,
    substr: string,
    prefixSuffixTable: number[],
): number {
    // pointer i used within the bigstr. i always moves forward
    let i = 0;
    /* pointer j used within the substr. j moves forward until
    match is not found and then moves backward intelligently */
    let j = 0;

    while (i < bigstr.length && j < substr.length) {
        if (bigstr[i] === substr[j]) {
            i += 1;
            j += 1;
        } else if (j !== 0) {
            j = prefixSuffixTable[j - 1] + 1;
        } else {
            i += 1;
        }
    }
    return j === substr.length ? i - j : -1;
}

/**
 * This algorithm represents the Knuth-Morris-Pratt(KMP)
 * string-searching algorithm. This algorithm is used to search
 * for occurrences of a word X within another word Y.
 *
 * Time:
 *     O(n+m) best/avg/worst
 * Space:
 *     O(m) best/avg/worst
 *
 * Where n is the length of string1, m is the length of the string
 * your looking for in string 1.
 *
 * @param {string} bigstr String we are looking for the substr pattern in
 * @param {string} substr Pattern we are looking for in the big string
 * @returns {number} Number indicating where the substr starts within the bigstr
 * . If not found, function will return -1.
 */
function knuthMorrisPratt(bigstr: string, substr: string): number {
    if (substr == null || substr.length === 0) {
        return -1;
    }
    const prefixSuffixTable = makePrefixSuffixTable(substr);
    return match(bigstr, substr, prefixSuffixTable);
}
export default knuthMorrisPratt;
